### PR TITLE
[SPARK-33272][SQL] prune the attributes mapping in QueryPlan.transformUpWithNewOutput

### PR DESCRIPTION
**Pull request by cloud-fan**
### What changes were proposed in this pull request?

For complex query plans, `QueryPlan.transformUpWithNewOutput` will keep accumulating the attributes mapping to be propagated, which may hurt performance. This PR prunes the attributes mapping before propagating.

### Why are the changes needed?

A simple perf improvement.

### Does this PR introduce _any_ user-facing change?

No

### How was this patch tested?

existing tests

### CODE REVIEW

**Comment by cloud-fan** [comment]
cc @maropu @viirya

**Review by viirya — APPROVED** [review]

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/QueryPlan.scala [hunk 1]
```diff
@@ -180,10 +180,14 @@ abstract class QueryPlan[PlanType <: QueryPlan[PlanType]] extends TreeNode[PlanT
    *             rewrite attribute references in the parent nodes.
    * @param skipCond a boolean condition to indicate if we can skip transforming a plan node to save
    *                 time.
+   * @param canGetOutput a boolean condition to indicate if we can get the output of a plan node
+   *                     to prune the attributes mapping to be propagated. The default value is true
+   *                     as only unresolved logical plan can't get output.
```
Comment (viirya) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/QueryPlan.scala:185]: nit: Is this needed to be parameterized? Seems we only need check unresolved logical plan.

Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/QueryPlan.scala:185]: Do you mean something like
```
val canGetOutput = plan match {
  case l: LogicalPlan => l.resolved
  case _ => true
}
```
then we can remove `skipCond` as well by matching `AnalysisHelper`.

Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/QueryPlan.scala:185]: One advantage of the current way is we can skip the check for optimizer rules (use the default value true).

Comment (viirya) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/QueryPlan.scala:185]: Oh, I see. Make sense.

**Review by maropu — APPROVED** [review]
Good improvement! LGTM

**Comment by maropu** [comment]
Thanks@ Merged to master.

**Comment by gaoyajun02** [comment]
hi @cloud-fan, can you help to open backport PRs for 3.0.2? thanks!

I found that prunes the attributes mapping is not only performance optimized, it also circumvents the bug of rewriting attributes for some complex queries, e.g.  https://issues.apache.org/jira/browse/SPARK-36815

**Comment by cloud-fan** [comment]
@gaoyajun02 can you cherry-pick this commit and open the backport PR? I'm quite busy this week and I can do the backport next week if you are not able to do it.

**Pull request closed without merging**